### ISSUE TITLE
Sealed Traits Fail to Serialize in Scala 2.11.8

### ISSUE THREAD

**Issue by abhsrivastava** [circe/circe#829]
I have the following code

build.sbt

```
name := "CirceEncoderTest"
version := "0.1"
scalaVersion := "2.12.4"
libraryDependencies ++= Seq(
   "io.circe" %% "circe-core" % "0.9.1",
   "io.circe" %% "circe-generic" % "0.9.1",
   "io.circe" %% "circe-parser" % "0.9.1"
)
```

MyModels.scala

```
sealed trait Foo
final case class Bar(name: String) extends Foo
```

MyApplication.scala

```
import io.circe.syntax._, io.circe._, io.circe.generic.auto._
object MyApplication extends App {
   val x : Foo = Bar("Foo")
   println(x.asJson.toString)
}

```

If I run this code. it works perfectly and produces the output

```
{
  "Bar" : {
    "name" : "Foo"
  }
}
```

But now if I change the scala version in build.sbt

```
name := "CirceEncoderTest"
version := "0.1"
scalaVersion := "2.11.8"
libraryDependencies ++= Seq(
   "io.circe" %% "circe-core" % "0.9.1",
   "io.circe" %% "circe-generic" % "0.9.1",
   "io.circe" %% "circe-parser" % "0.9.1"
)
addCompilerPlugin("org.scalamacros" % "paradise" % "2.1.0" cross CrossVersion.full)
```

Then the same code fails to run with error

`[error] MyApplication.scala:5:14: could not find implicit value for parameter encoder: io.circe.Encoder[Foo]
`

I tried the standard workarounds of using @JsonCodec before the sealed trait and adding an empty Object Foo after the case class. both didn't solve the problem.

Talked to @liff at gitter channel and he asked me to remove the io.circe.generic.auto._ from the MyApplication as it may be conflicting with @JsonCodec but that also did not solve the issue.

I have checked in my sample code here

https://github.com/abhsrivastava/CirceEncoderTest

The branch "works" contains the working code

https://github.com/abhsrivastava/CirceEncoderTest/tree/works

**Comment by pfcoperez** [comment]
I could make it work by moving your models definitions to `MyApplication` (outside `MyApplication` object):

```scala
import io.circe.generic.auto._, io.circe.syntax._                                                       
                                                                                                        
sealed trait Foo                                                                                        
final case class Bar(name: String) extends Foo                                                          
                                                                                                        
object MyApplication extends App {                                                                      
   val x : Foo = Bar("Foo")                                                                             
   println(x.asJson.toString)                                                                           
}  
```

Could it be related to the nested class problems https://github.com/circe/circe/issues/69 ? Maybe some corner case in Scala 2.11?

**Comment by travisbrown** [comment]
This is almost certainly a case of SI-7046, which is (partially) fixed by 2.11.12 (I can't remember exactly which 2.11 version the fix showed up in, but it was after 2.11.8).

I confirmed on my laptop that your code fails to compile on 2.11.8 and succeeds on 2.11.12. Unfortunately we can't support old patch Scala versions, so I'm going to close this issue—please reopen if you can reproduce the issue on 2.11.12.
